### PR TITLE
CUDA: Fix dropping of kernels by nvjitlink, by implementing the used list

### PR DESCRIPTION
**Pull request by gmarkall**
The compiler or linker are at liberty to drop kernels if they are not explicitly added to the [`@"llvm.used"`](https://releases.llvm.org/7.0.1/docs/LangRef.html#the-llvm-used-global-variable) global variable. This has not been an issue so far because NVVM's behaviour has always been to preserve kernels; however, compiling / linking with nvJitLink causes anything not referenced in the used list (directly or indirectly) to be dropped.

When setting a function as a CUDA kernel, a used list is now created and the kernel is added to it - this assumes one kernel per module, which is true for Numba. Testing ensures that we correctly construct the used list, and the existing test IR in `test_nvvm_driver` is updated to contain a used list.

As part of this fix, some related changes were needed:

- A tidy-up of `test_nvvm_driver` to remove some obsolete / useless test infrastructure.
- Fixing missing NVVM dumps for CUDA kernel wrappers - this is not strictly essential, but it was a bug, and fixing it enables one to actually observe the used list when dumping LLVM IR for a CUDA kernel.

### CODE REVIEW

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
cc @leopoldcambier

**Comment by gmarkall** [comment]
gpuci run tests

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @gmarkall. One minor query to resolve noted in line else looks good.

File: numba/cuda/cudadrv/nvvm.py [hunk 1]
```diff
@@ -653,19 +653,43 @@ def llvm140_to_70_ir(ir):
...
-    lfunc.attributes.discard('noinline')
+    # Create the used list
+    ptrty = ir.IntType(8).as_pointer()
+    usedty = ir.ArrayType(ptrty, 1)
+
+    fnptr = function.bitcast(ptrty)
+
+    llvm_used = ir.GlobalVariable(module, usedty, 'llvm.used')
+    llvm_used.linkage = 'appending'
+    llvm_used.section = 'llvm.metadata'
```
[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/cuda/cudadrv/nvvm.py:688]: Yes (from the PR description):

> When setting a function as a CUDA kernel, a used list is now created and the kernel is added to it - this assumes one kernel per module, which is true for Numba.

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/cuda/cudadrv/nvvm.py:688]: Thanks, I knew I'd seen that somewhere but couldn't recall where!

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch. Skipping build farm run as the tests are mostly checking LLVM module strings and `gpuCI` would pick up any issues in the code changes.

**Comment by esc** [comment]
@gmarkall this is now in conflict with `main`, can you resolve?

**Comment by gmarkall** [comment]
gpuCI is having an issue at the moment, marking this as "Waiting for CI" as a reminder.

**Comment by gmarkall** [comment]
gpuci run tests

**Review by stuartarchibald — APPROVED** [review]
Re-approving, merge looks good. Thanks @gmarkall.

**Pull request merged**